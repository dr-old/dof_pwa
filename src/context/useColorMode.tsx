import * as React from "react";
import { ThemeProvider, createTheme } from "@mui/material/styles";
import CssBaseline from "@mui/material/CssBaseline";
import { colors } from "../utils/colors";

interface ColorModeContextType {
  toggleColorMode: () => void;
}

const ColorModeContext = React.createContext<ColorModeContextType | undefined>(
  undefined
);

export const useColorMode = (): ColorModeContextType => {
  const context = React.useContext(ColorModeContext);
  if (!context) {
    throw new Error("useColorMode must be used within a ColorModeProvider");
  }
  return context;
};

interface ColorModeProviderProps {
  children: React.ReactNode;
}

export const ColorModeProvider: React.FC<ColorModeProviderProps> = ({
  children,
}) => {
  const [mode, setMode] = React.useState<"light" | "dark">("light");

  const colorMode = React.useMemo(
    () => ({
      toggleColorMode: () => {
        setMode((prevMode) => (prevMode === "light" ? "dark" : "light"));
      },
    }),
    []
  );

  const theme = React.useMemo(
    () =>
      createTheme({
        palette: {
          mode,
          ...(mode === "light" ? colors.light : colors.dark),
        },
        typography: {
          fontFamily: ['"Poppins"', "sans-serif"].join(","),
        },
      }),
    [mode]
  );

  return (
    <ColorModeContext.Provider value={colorMode}>
      <ThemeProvider theme={theme}>
        <CssBaseline />
        {children}
      </ThemeProvider>
    </ColorModeContext.Provider>
  );
};
