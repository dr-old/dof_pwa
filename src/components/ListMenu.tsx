import React from "react";
import List from "@mui/material/List";
import ListItem from "@mui/material/ListItem";
import DashboardRounded from "@mui/icons-material/DashboardRounded";
import TextSnippetRounded from "@mui/icons-material/TextSnippetRounded";
import EventRounded from "@mui/icons-material/EventRounded";
import TuneRounded from "@mui/icons-material/TuneRounded";
import {
  Box,
  ListItemButton,
  Avatar,
  Typography,
  Button,
  useTheme,
  IconButton,
} from "@mui/material";
import { styled } from "@mui/system";
import {
  CloseRounded,
  EditRounded,
  ExitToAppRounded,
} from "@mui/icons-material";
import { useAuth } from "../context/useAuth";
import { useLocation, useNavigate } from "react-router-dom";
import ToggleTheme from "./ToggleTheme";
import AlertDialog from "./AlertDialog";

// Styled component for the ListItemButton to make it rounded
const RoundedListItemButton = styled(ListItemButton)(({ theme }) => ({
  borderRadius: theme.shape.borderRadius * 3,
  transition: "background-color 0.3s ease",
  "&:hover": {
    opacity: 1,
    backgroundColor: theme.palette.action.hover,
  },
}));

interface ListMenuProps {
  isMenuOpen: boolean;
  isMobile: boolean;
  toggleMenu: () => void;
}

const ListMenu: React.FC<ListMenuProps> = ({
  isMobile,
  isMenuOpen,
  toggleMenu,
}) => {
  const { state } = useLocation();
  const { logOut, user } = useAuth();
  const theme = useTheme();
  const navigate = useNavigate();
  const [open, setOpen] = React.useState(false); // Controls the AlertDialog

  const handleOpen = () => {
    setOpen(true);
  };

  const handleClose = () => {
    setOpen(false);
  };

  const handleLogout = () => {
    setOpen(false);
    logOut();
    navigate("/login", { replace: true });
  };

  // Define the menu items
  const menu = [
    {
      title: "Dashboard",
      onClick: () => {
        navigate("/dashboard", { state: { title: "Dashboard" } });
      },
      icon: <DashboardRounded fontSize="small" />,
    },
    {
      title: "Activity",
      onClick: () => {
        navigate("/activity", { state: { title: "Activity" } });
      },
      icon: <TextSnippetRounded fontSize="small" />,
    },
    {
      title: "Schedule",
      onClick: () => {
        navigate("/schedule", { state: { title: "Schedule" } });
      },
      icon: <EventRounded fontSize="small" />,
    },
    {
      title: "Settings",
      onClick: () => {
        navigate("/settings", { state: { title: "Settings" } });
      },
      icon: <TuneRounded fontSize="small" />,
    },
    {
      title: "Log Out",
      onClick: () => handleOpen(),
      icon: <ExitToAppRounded fontSize="small" />,
    },
  ];

  return (
    <React.Fragment>
      {/* Sliding Menu */}
      <Box
        sx={[
          {
            height: "100%",
            minWidth: "250px",
            px: 2,
            backgroundColor: theme.palette.background.default,
          },
          isMobile
            ? {
                pb: 2,
                position: "fixed",
                top: 0,
                right: 0, // Changed from left to right
                transform: isMenuOpen ? "translateX(0)" : "translateX(100%)", // Updated translateX
                transition: "transform 0.3s ease-in-out",
                boxShadow: isMenuOpen ? theme.shadows[5] : "none",
                zIndex: 1200, // Ensure it stays on top
                overflowX: "auto",
              }
            : {
                my: { xs: 1, sm: 2, md: 3, lg: 4 },
              },
        ]}>
        {/* User info section */}
        <Box
          sx={{
            display: "flex",
            flexDirection: "column",
            alignItems: "center",
            padding: 2,
          }}>
          {isMobile && (
            <IconButton onClick={toggleMenu} sx={{ alignSelf: "flex-start" }}>
              <CloseRounded fontSize="small" />
            </IconButton>
          )}
          <Typography
            variant="h5"
            color={
              theme.palette.mode === "dark" ? "warning.main" : "primary.main"
            }
            fontWeight={600}
            sx={{ mb: 3, mt: 2 }}>
            DOF
          </Typography>
          <Avatar
            src={user?.photo || "https://i.pravatar.cc/150?img=2"}
            alt={user?.fullname}
            sx={{ width: 80, height: 80, marginBottom: 2 }}
          />
          <Typography variant="caption" sx={{ mb: 1 }}>
            {user?.fullname}
          </Typography>
          {/* Replace with actual name */}
          <Button
            variant="outlined"
            size="small"
            startIcon={<EditRounded fontSize="small" />}
            sx={{
              marginTop: 1,
              fontSize: "0.5rem",
              borderRadius: 20,
              textTransform: "none",
              paddingX: 2,
            }}>
            <Typography variant="caption">Edit Profile</Typography>
          </Button>
        </Box>

        {/* Menu items */}
        <List>
          {menu.map((item, index) => (
            <ListItem key={index}>
              <RoundedListItemButton
                sx={{
                  py: 2,
                  opacity: item.title === state.title ? 1 : 0.6,
                  backgroundColor:
                    item.title === state.title
                      ? theme.palette.action.hover
                      : "transparent",
                }}
                onClick={item.onClick}>
                {item.icon}
                <Typography variant="caption" paddingLeft={1.5}>
                  {item.title}
                </Typography>
              </RoundedListItemButton>
            </ListItem>
          ))}
        </List>
        <Box
          sx={{
            display: "flex",
            flexDirection: "column",
            alignItems: "center",
          }}>
          <ToggleTheme />
        </Box>
      </Box>

      <AlertDialog
        open={open}
        onClose={handleClose}
        title="Confirm Logout"
        content="Are you sure you want to log out?"
        onConfirm={handleLogout}
        confirmText="Logout"
        cancelText="Cancel"
      />
    </React.Fragment>
  );
};

export default ListMenu;
