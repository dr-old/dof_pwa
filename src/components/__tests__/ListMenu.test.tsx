import React, { ReactNode } from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import ListMenu from "../ListMenu";
import { ThemeProvider, createTheme } from "@mui/material/styles";
import { MemoryRouter } from "react-router-dom";
import { useAuth } from "../../context/useAuth";
import { useLocation, useNavigate } from "react-router-dom";
import { ColorModeProvider } from "../../context/useColorMode";

// Mock the useAuth hook
jest.mock("../../context/useAuth", () => ({
  useAuth: jest.fn(),
}));

// Mock the useLocation and useNavigate hooks
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useLocation: jest.fn(),
  useNavigate: jest.fn(),
}));

interface MockColorModeProviderProps {
  children: ReactNode;
}

// Mock ColorModeProvider, if required
const MockColorModeProvider: React.FC<MockColorModeProviderProps> = ({
  children,
}) => <ColorModeProvider>{children}</ColorModeProvider>;

describe("ListMenu Component", () => {
  const theme = createTheme();
  const mockToggleMenu = jest.fn();
  const mockNavigate = jest.fn();

  beforeEach(() => {
    (useAuth as jest.Mock).mockReturnValue({
      user: {
        fullname: "John Doe",
        photo: "https://i.pravatar.cc/150?img=2",
      },
      logOut: jest.fn(),
    });

    (useLocation as jest.Mock).mockReturnValue({
      pathname: "/dashboard",
      state: { title: "Dashboard" },
    });

    (useNavigate as jest.Mock).mockReturnValue(mockNavigate);
  });

  const renderComponent = (isMenuOpen: boolean, isMobile: boolean) =>
    render(
      <ThemeProvider theme={theme}>
        <MemoryRouter>
          <MockColorModeProvider>
            <ListMenu
              isMenuOpen={isMenuOpen}
              isMobile={isMobile}
              toggleMenu={mockToggleMenu}
            />
          </MockColorModeProvider>
        </MemoryRouter>
      </ThemeProvider>
    );

  it("renders the component", () => {
    renderComponent(true, false);
    expect(screen.getByText("John Doe")).toBeInTheDocument();
    // Add additional assertions as needed
  });

  it("matches the snapshot", () => {
    const { asFragment } = renderComponent(true, false);
    expect(asFragment()).toMatchSnapshot();
  });
});
