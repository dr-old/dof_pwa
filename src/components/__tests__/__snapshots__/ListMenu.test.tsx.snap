// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListMenu Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-18xot1z"
  >
    <div
      class="MuiBox-root css-162is1w"
    >
      <h5
        class="MuiTypography-root MuiTypography-h5 css-xmu7xk-MuiTypography-root"
      >
        DOF
      </h5>
      <div
        class="MuiAvatar-root MuiAvatar-circular css-5kg9jr-MuiAvatar-root"
      >
        <img
          alt="John Doe"
          class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
          src="https://i.pravatar.cc/150?img=2"
        />
      </div>
      <span
        class="MuiTypography-root MuiTypography-caption css-93il58-MuiTypography-root"
      >
        John Doe
      </span>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorPrimary css-e3rzpn-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeSmall css-y6rp3m-MuiButton-startIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
            data-testid="EditRoundedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 17.46v3.04c0 .28.22.5.5.5h3.04c.13 0 .26-.05.35-.15L17.81 9.94l-3.75-3.75L3.15 17.1c-.1.1-.15.22-.15.36M20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
            />
          </svg>
        </span>
        <span
          class="MuiTypography-root MuiTypography-caption css-1b8u2zs-MuiTypography-root"
        >
          Edit Profile
        </span>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <ul
      class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
    >
      <li
        class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-12h1lx2-MuiListItem-root"
      >
        <div
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-mmvunt-MuiButtonBase-root-MuiListItemButton-root"
          role="button"
          tabindex="0"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
            data-testid="DashboardRoundedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M4 13h6c.55 0 1-.45 1-1V4c0-.55-.45-1-1-1H4c-.55 0-1 .45-1 1v8c0 .55.45 1 1 1m0 8h6c.55 0 1-.45 1-1v-4c0-.55-.45-1-1-1H4c-.55 0-1 .45-1 1v4c0 .55.45 1 1 1m10 0h6c.55 0 1-.45 1-1v-8c0-.55-.45-1-1-1h-6c-.55 0-1 .45-1 1v8c0 .55.45 1 1 1M13 4v4c0 .55.45 1 1 1h6c.55 0 1-.45 1-1V4c0-.55-.45-1-1-1h-6c-.55 0-1 .45-1 1"
            />
          </svg>
          <span
            class="MuiTypography-root MuiTypography-caption css-1b1yxjp-MuiTypography-root"
          >
            Dashboard
          </span>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </div>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-12h1lx2-MuiListItem-root"
      >
        <div
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-qw940o-MuiButtonBase-root-MuiListItemButton-root"
          role="button"
          tabindex="0"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
            data-testid="TextSnippetRoundedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M14.17 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V9.83c0-.53-.21-1.04-.59-1.41l-4.83-4.83c-.37-.38-.88-.59-1.41-.59M8 15h8c.55 0 1 .45 1 1s-.45 1-1 1H8c-.55 0-1-.45-1-1s.45-1 1-1m0-4h8c.55 0 1 .45 1 1s-.45 1-1 1H8c-.55 0-1-.45-1-1s.45-1 1-1m0-4h5c.55 0 1 .45 1 1s-.45 1-1 1H8c-.55 0-1-.45-1-1s.45-1 1-1"
            />
          </svg>
          <span
            class="MuiTypography-root MuiTypography-caption css-1b1yxjp-MuiTypography-root"
          >
            Activity
          </span>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </div>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-12h1lx2-MuiListItem-root"
      >
        <div
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-qw940o-MuiButtonBase-root-MuiListItemButton-root"
          role="button"
          tabindex="0"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
            data-testid="EventRoundedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16 13h-3c-.55 0-1 .45-1 1v3c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-3c0-.55-.45-1-1-1m0-10v1H8V3c0-.55-.45-1-1-1s-1 .45-1 1v1H5c-1.11 0-1.99.9-1.99 2L3 20c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2h-1V3c0-.55-.45-1-1-1s-1 .45-1 1m2 17H6c-.55 0-1-.45-1-1V9h14v10c0 .55-.45 1-1 1"
            />
          </svg>
          <span
            class="MuiTypography-root MuiTypography-caption css-1b1yxjp-MuiTypography-root"
          >
            Schedule
          </span>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </div>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-12h1lx2-MuiListItem-root"
      >
        <div
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-qw940o-MuiButtonBase-root-MuiListItemButton-root"
          role="button"
          tabindex="0"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
            data-testid="TuneRoundedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 18c0 .55.45 1 1 1h5v-2H4c-.55 0-1 .45-1 1M3 6c0 .55.45 1 1 1h9V5H4c-.55 0-1 .45-1 1m10 14v-1h7c.55 0 1-.45 1-1s-.45-1-1-1h-7v-1c0-.55-.45-1-1-1s-1 .45-1 1v4c0 .55.45 1 1 1s1-.45 1-1M7 10v1H4c-.55 0-1 .45-1 1s.45 1 1 1h3v1c0 .55.45 1 1 1s1-.45 1-1v-4c0-.55-.45-1-1-1s-1 .45-1 1m14 2c0-.55-.45-1-1-1h-9v2h9c.55 0 1-.45 1-1m-5-3c.55 0 1-.45 1-1V7h3c.55 0 1-.45 1-1s-.45-1-1-1h-3V4c0-.55-.45-1-1-1s-1 .45-1 1v4c0 .55.45 1 1 1"
            />
          </svg>
          <span
            class="MuiTypography-root MuiTypography-caption css-1b1yxjp-MuiTypography-root"
          >
            Settings
          </span>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </div>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-12h1lx2-MuiListItem-root"
      >
        <div
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-qw940o-MuiButtonBase-root-MuiListItemButton-root"
          role="button"
          tabindex="0"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
            data-testid="ExitToAppRoundedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10.79 16.29c.39.39 1.02.39 1.41 0l3.59-3.59c.39-.39.39-1.02 0-1.41L12.2 7.7a.9959.9959 0 0 0-1.41 0c-.39.39-.39 1.02 0 1.41L12.67 11H4c-.55 0-1 .45-1 1s.45 1 1 1h8.67l-1.88 1.88c-.39.39-.38 1.03 0 1.41M19 3H5c-1.11 0-2 .9-2 2v3c0 .55.45 1 1 1s1-.45 1-1V6c0-.55.45-1 1-1h12c.55 0 1 .45 1 1v12c0 .55-.45 1-1 1H6c-.55 0-1-.45-1-1v-2c0-.55-.45-1-1-1s-1 .45-1 1v3c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2"
            />
          </svg>
          <span
            class="MuiTypography-root MuiTypography-caption css-1b1yxjp-MuiTypography-root"
          >
            Log Out
          </span>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </div>
      </li>
    </ul>
    <div
      class="MuiBox-root css-dvxtzn"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-colorPrimary css-1ssabq6-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeSmall css-y6rp3m-MuiButton-startIcon"
        >
          <div
            class="MuiBox-root css-a4pqvs"
            data-testid="brightness4-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
              data-testid="Brightness4RoundedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M22.6 11.29 20 8.69V5c0-.55-.45-1-1-1h-3.69l-2.6-2.6a.9959.9959 0 0 0-1.41 0L8.69 4H5c-.55 0-1 .45-1 1v3.69l-2.6 2.6c-.39.39-.39 1.02 0 1.41L4 15.3V19c0 .55.45 1 1 1h3.69l2.6 2.6c.39.39 1.02.39 1.41 0l2.6-2.6H19c.55 0 1-.45 1-1v-3.69l2.6-2.6c.39-.39.39-1.03 0-1.42m-4.68 1.69c-.34 2.12-1.85 3.94-3.88 4.66-1.21.43-2.41.45-3.5.18-.41-.1-.48-.65-.13-.9C11.98 15.84 13 14.04 13 12s-1.02-3.84-2.58-4.92c-.35-.24-.29-.79.13-.9 1.09-.27 2.29-.25 3.5.18 2.02.72 3.54 2.54 3.88 4.66.05.33.07.66.07.98-.01.32-.03.65-.08.98"
              />
            </svg>
          </div>
        </span>
        <span
          class="MuiTypography-root MuiTypography-caption css-1569pzt-MuiTypography-root"
        >
          light mode
        </span>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
